### PR TITLE
Build: Move java home checks to pre-execution phase

### PR DESCRIPTION
**Pull request by rjernst**
This commit moves the checks on JAVAX_HOME (where X is the java version
number) existing to the end of gradle's configuration phase, and based
on whether the tasks needing the java home are configured to execute.

relates #29519

### CODE REVIEW

**Comment by rjernst** [comment]
With this change, the output for running `./gradlew :qa:full-cluster-restart:bwcTest` on master now looks like:

```
* What went wrong:
Failed to notify task execution graph listener.
> JAVA9_HOME required to run task :distribution:bwc:next-bugfix-snapshot:buildBwcVersion
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.0.0#oldClusterTestCluster#node0.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.0.0#oldClusterTestCluster#node1.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.0.1#oldClusterTestCluster#node0.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.0.1#oldClusterTestCluster#node1.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.1.0#oldClusterTestCluster#node0.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.1.0#oldClusterTestCluster#node1.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.1.1#oldClusterTestCluster#node0.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.1.1#oldClusterTestCluster#node1.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.1.2#oldClusterTestCluster#node0.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.1.2#oldClusterTestCluster#node1.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.1.3#oldClusterTestCluster#node0.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.1.3#oldClusterTestCluster#node1.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.1.4#oldClusterTestCluster#node0.start
> JAVA8_HOME required to run task :qa:full-cluster-restart:v6.1.4#oldClusterTestCluster#node1.start
> JAVA9_HOME required to run task :qa:full-cluster-restart:v6.2.0#oldClusterTestCluster#node0.start
> JAVA9_HOME required to run task :qa:full-cluster-restart:v6.2.0#oldClusterTestCluster#node1.start
> JAVA9_HOME required to run task :qa:full-cluster-restart:v6.2.1#oldClusterTestCluster#node0.start
> JAVA9_HOME required to run task :qa:full-cluster-restart:v6.2.1#oldClusterTestCluster#node1.start
> JAVA9_HOME required to run task :qa:full-cluster-restart:v6.2.2#oldClusterTestCluster#node0.start
> JAVA9_HOME required to run task :qa:full-cluster-restart:v6.2.2#oldClusterTestCluster#node1.start
> JAVA9_HOME required to run task :qa:full-cluster-restart:v6.2.3#oldClusterTestCluster#node0.start
> JAVA9_HOME required to run task :qa:full-cluster-restart:v6.2.3#oldClusterTestCluster#node1.start
> JAVA9_HOME required to run task :qa:full-cluster-restart:v6.2.4-SNAPSHOT#oldClusterTestCluster#node0.start
> JAVA9_HOME required to run task :qa:full-cluster-restart:v6.2.4-SNAPSHOT#oldClusterTestCluster#node1.start
```

And for `./gradlew :qa:reindex-from-old:check`:

```
* What went wrong:
Failed to notify task execution graph listener.
> JAVA7_HOME required to run task :qa:reindex-from-old:oldEs2Fixture
> JAVA7_HOME required to run task :qa:reindex-from-old:oldEs1Fixture
> JAVA7_HOME required to run task :qa:reindex-from-old:oldEs090Fixture
```

**Review by nik9000 — APPROVED** [review]

**Review by jasontedor — APPROVED** [review]
LGTM. Great change.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy [hunk 1]
```diff
@@ -221,20 +222,20 @@ class BuildPlugin implements Plugin<Project> {
...
-     * @param version the version of Java home to obtain
-     * @param message the exception message if Java home for the specified version is not configured
-     * @return Java home for the specified version
-     * @throws GradleException if Java home for the specified version is not configured
-     */
-    static String getJavaHome(final Project project, final int version, final String message) {
-        if (project.javaVersions.get(version) == null) {
-            throw new GradleException(message)
+    /** Add a check before gradle execution phase which ensures java home for the given java version is set. */
+    static void requireJavaHome(Project project, Task task, int version) {
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: Yep, changed in a98747a.

**Comment by rjernst** [comment]
Thanks for the reviews. Just FYI, I iterated on the output a little bit, and now have the following:

```
* What went wrong:
JAVA8_HOME required to run tasks:
  :qa:full-cluster-restart:v6.0.0#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.0.0#oldClusterTestCluster#node1.start
  :qa:full-cluster-restart:v6.0.1#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.0.1#oldClusterTestCluster#node1.start
  :qa:full-cluster-restart:v6.1.0#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.1.0#oldClusterTestCluster#node1.start
  :qa:full-cluster-restart:v6.1.1#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.1.1#oldClusterTestCluster#node1.start
  :qa:full-cluster-restart:v6.1.2#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.1.2#oldClusterTestCluster#node1.start
  :qa:full-cluster-restart:v6.1.3#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.1.3#oldClusterTestCluster#node1.start
  :qa:full-cluster-restart:v6.1.4#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.1.4#oldClusterTestCluster#node1.start
JAVA9_HOME required to run tasks:
  :distribution:bwc:next-bugfix-snapshot:buildBwcVersion
  :qa:full-cluster-restart:v6.2.0#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.2.0#oldClusterTestCluster#node1.start
  :qa:full-cluster-restart:v6.2.1#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.2.1#oldClusterTestCluster#node1.start
  :qa:full-cluster-restart:v6.2.2#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.2.2#oldClusterTestCluster#node1.start
  :qa:full-cluster-restart:v6.2.3#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.2.3#oldClusterTestCluster#node1.start
  :qa:full-cluster-restart:v6.2.4-SNAPSHOT#oldClusterTestCluster#node0.start
  :qa:full-cluster-restart:v6.2.4-SNAPSHOT#oldClusterTestCluster#node1.start```

**Comment by nik9000** [comment]
> I iterated on the output a little bit, and now have the following

I like it better!

**Comment by rjernst** [comment]
@elasticmachine test this please

**Pull request merged**